### PR TITLE
Add a crapton of missing trait impls and docs

### PR DESCRIPTION
**Pull request by Gankra**
This is almost entirely copy-pasting from std. I also fleshed out a bunch of docs in Drain. This is a yakshave to prep for adding splice support.

### CODE REVIEW

**Comment by Gankra** [comment]
bonus: actually handling isize::MAX stuff

File: src/lib.rs [hunk 1]
```diff
@@ -326,22 +331,26 @@ extern "C" {
     static EMPTY_HEADER: Header;
 }
 
-// TODO: overflow checks everywhere
-
 // Utils for computing layouts of allocations
 
 fn alloc_size<T>(cap: usize) -> usize {
```
Comment (nnethercote) [inline src/lib.rs:336]: Add a doc comment to `alloc_size` explaining that it can panic?

File: src/lib.rs [hunk 2]
```diff
@@ -439,10 +448,67 @@ macro_rules! thin_vec {
...
+    /// reallocating. This method is allowed to allocate for more elements than
+    /// `capacity`. If `capacity` is 0, the vector will not allocate.
+    ///
+    /// It is important to note that although the returned vector has the
+    /// minimum *capacity* specified, the vector will have a zero *length*.
+    ///
+    /// If it is important to know the exact allocated capacity of a `ThinVec`,
+    /// always use the [`capacity`] method after construction.
+    ///
+    /// **NOTE**: unlike Vec, ThinVec **MUST** allocate to keep track of non-zero
```
Comment (nnethercote) [inline src/lib.rs:470]: Nit: Vec and ThinVec here don't have backticks. I'm not sure how hard you're trying to get every one.

File: src/lib.rs [hunk 3]
```diff
@@ -525,16 +591,134 @@ impl<T> ThinVec<T> {
...
+    ///
+    /// While the following example is sound, there is a memory leak since
+    /// the inner vectors were not freed prior to the `set_len` call:
+    ///
+    /// ```no_run
+    /// use thin_vec::thin_vec;
+    ///
+    /// let mut vec = thin_vec![thin_vec![1, 0, 0],
+    ///                    thin_vec![0, 1, 0],
+    ///                    thin_vec![0, 0, 1]];
```
Comment (nnethercote) [inline src/lib.rs:711]: Formatting on this line could be nicer.

File: src/lib.rs [hunk 4]
```diff
@@ -1415,16 +1886,258 @@ impl<T> FromIterator<T> for ThinVec<T> {
...
+    fn from(s: [T; N]) -> ThinVec<T> {
+        std::iter::IntoIterator::into_iter(s).collect()
+    }
+}
+
+impl<T> From<Box<[T]>> for ThinVec<T> {
+    /// Convert a boxed slice into a vector by transferring ownership of
+    /// the existing heap allocation.
+    ///
+    /// **NOTE:** unlike std, this must reallocate to change the layout!
```
Comment (nnethercote) [inline src/lib.rs:1939]: `std`? (i.e. with backticks)

File: src/lib.rs [hunk 5]
```diff
@@ -1415,16 +1886,258 @@ impl<T> FromIterator<T> for ThinVec<T> {
...
+    /// assert_eq!(ThinVec::from(b), thin_vec![1, 2, 3]);
+    /// ```
+    fn from(s: Box<[T]>) -> Self {
+        // Can just lean on the fact that `Box<[T]>` -> `Vec<T>` is Free.
+        Vec::from(s).into_iter().collect()
+    }
+}
+
+impl<T> From<Vec<T>> for ThinVec<T> {
+    /// Convert a std::Vec into a ThinVec.
```
Comment (nnethercote) [inline src/lib.rs:1956]: Backticks?

Comment (nnethercote) [reply src/lib.rs:1956]: More cases below, I won't point them all out, a bit of grepping will find them if you care to change them.

File: src/lib.rs [hunk 6]
```diff
@@ -1415,16 +1886,258 @@ impl<T> FromIterator<T> for ThinVec<T> {
...
+    /// # Examples
+    ///
+    /// ```
+    /// use thin_vec::{ThinVec, thin_vec};
+    ///
+    /// let b: Vec<i32> = vec![1, 2, 3];
+    /// assert_eq!(ThinVec::from(b), thin_vec![1, 2, 3]);
+    /// ```
+    fn from(s: Vec<T>) -> Self {
+        // Can just lean on the fact that `Box<[T]>` -> `Vec<T>` is Free.
```
Comment (nnethercote) [inline src/lib.rs:1969]: Why is `Box<[T]>` relevant here?

File: src/lib.rs [hunk 7]
```diff
@@ -1415,16 +1886,258 @@ impl<T> FromIterator<T> for ThinVec<T> {
...
+    /// # Examples
+    ///
+    /// ```
+    /// use thin_vec::{ThinVec, thin_vec};
+    ///
+    /// let b: ThinVec<i32> = thin_vec![1, 2, 3];
+    /// assert_eq!(Vec::from(b), vec![1, 2, 3]);
+    /// ```
+    fn from(s: ThinVec<T>) -> Self {
+        // Can just lean on the fact that `Box<[T]>` -> `Vec<T>` is Free.
```
Comment (nnethercote) [inline src/lib.rs:1988]: ditto

File: src/lib.rs [hunk 8]
```diff
@@ -1477,6 +2197,126 @@ impl<T> Drop for IntoIter<T> {
...
+    //
+    // All this requires is moving the tail back to the prefix (stored in `len`)
+    // and setting `len` to `len + tail_len` to undo the leak amplification.
+    /// An iterator over the elements we're removing.
+    ///
+    /// As we go we'll be `read`ing out of the mutable refs yielded by this.
+    /// It's ok to use IterMut here because it promises to only take mutable
+    /// refs to the parts we haven't yielded yet.
+    ///
+    /// A downside of this (and the *mut below is that it makes this Invariant, when
```
Comment (nnethercote) [inline src/lib.rs:2304]: This sentence reads strangely. Are the parentheses in the wrong place? Also, "Inductive" doesn't need to be capitalized?

File: src/lib.rs [hunk 9]
```diff
@@ -1477,6 +2197,126 @@ impl<T> Drop for IntoIter<T> {
...
+    /// technically it could be covariant?)
+    iter: IterMut<'a, T>,
+    /// The actual ThinVec, which we need to hold onto to undo the leak amplification
+    /// and backshift the tail into place. This should only be accessed when we're
+    /// completely done with the IterMut in the `drop` impl of this type (or miri will get mad).
+    ///
+    /// Since we set the `len` of this to be before `IterMut`, we can use that `len`
+    /// to retrieve the index of the start of the drain range later.
```
Comment (nnethercote) [inline src/lib.rs:2312]: Nice comment! I think you enjoyed writing that one :)

**Review by nnethercote — APPROVED** [review]
Looks good. A few minor suggestions. I admit I didn't read all the example code in the doc comments, though I did read the comment prose and the code in the tests.

**Comment by nnethercote** [comment]
Note: there's a clippy warning that might be worth using an attribute to placate. GitHub wouldn't let me comment on the specific line for some reason.

File: src/lib.rs (hunk 3 above)
Comment (Gankra) [reply src/lib.rs:711]: This is the formatting std had *shrug*

File: src/lib.rs [hunk 10]
```diff
@@ -326,24 +331,34 @@ extern "C" {
...
-    let header_size = mem::size_of::<Header>();
-    let elem_size = mem::size_of::<T>();
-    let padding = padding::<T>();
+    //
+    // We turn everything into isizes here so that we can catch isize::MAX overflow,
+    // we never want to allow allocations larger than that!
+    let cap = cap as isize;
+    let header_size = mem::size_of::<Header>() as isize;
+    let elem_size = mem::size_of::<T>() as isize;
+    let padding = padding::<T>() as isize;
```
Comment (eggyal) [inline src/lib.rs:349]: Apologies for dropping in belatedly, but I was just reviewing changes between v0.2.9 and v0.2.11 (for rust-lang/rust#108245) and spotted this...

Isn't there a risk that these casts (okay, less so the latter three but certainly the `cap` one) might silently overflow `isize::MAX` and result in a negative `isize` value?  Better to use `TryInto`, I think?

Comment (Gankra) [reply src/lib.rs:349]: ...yikes! good catch. 

https://github.com/Gankra/thin-vec/pull/45

**Pull request merged**